hifs.1
.TH HIFS 1 "November 1997" "Linux" "Handy Information For SysAdmins v 1.4"

.SH NAME
hifs, xhifs \- Handy Information For Sysadmins

.SH SYNOPSIS
\fB hifs \fR[-vhd] [--version] [--help] [--debug]
.sp 0
\fB xhifs \fR[-vhd] [--version] [--help] [--debug]

.SH DESCRIPTION
\fBHifs \fR is a tool for watching the state your system. As it name implies, 
it is especially usefull for system administrators, but it can be used by 
normal users, too. The special use for system administrators is the amount of
control you have within one window. You can see the most cpu intensive or
memory wasting processes and you can DO something about them. Hifs lets you
renice processes, write a message to a certain user or just kill the process.
Hifs can be installed setuid-root with the option to su to root in hifs. One
26x24 xterm should be enough to completely monitor a computer system. 

.SH LIST OF FEATURES
.TP
.B Process listing
Hifs shows the processes that use the most system resources. System 
resources can be chosen by the user to be one of: percentage CPU, resident 
set size or vsize.
.TP
.B Small size
Hifs uses only a 26x24 text window, which is especially usefull under X, using 
\fBxhifs\fR.
.TP
.B Correct misuse
The user kan write messages to users, kill or renice processes, if he/she has 
enough priviliges, of course.
.TP
.B Process information
Hifs can show the following extra info on each process: command line, 
username, pid, wchan and priority.
.TP
.B System statistics
Hifs shows the following system statistics: Total CPU states, free/used 
memory/swap, load averages, the number of real/total logins/xlogins.
.TP
.B Su-to-root option
Hifs can be compiled as a setuid root program to let the user su to root and 
drop these priviliges again from within hifs. This prevents you from having 
a root window open all the time.
.TP
.B Filesystem information
Hifs monitors mounted filesystems, and gives a sign when they are getting full.
.TP
.B User information
A maximum of two groups of eight people can be defined. These people are 
monitored for their presence on the system.

.SH COMMAND LINE OPTIONS
The following command line options are available.
.TP
.B -h, --help
Show help about command-line arguments.
.TP
.B -v, --version
Show version information.
.TP
.B -d, --debug
Set debug mode. In debug mode, hifs waits for the user to press a key when \
there was a warning.

.SH INTERACTIVE COMMANDS
Most commands in hifs are interactive. They are:
.TP
.B h, ? 
Show online help.
.TP
.B s, <SPACE>
Toggle the \fBsort\fR mode. The processes are sorted by this criterion. It 
can be one of: sort by cpu usage, sort by resident set size or sort by vsize.
.TP
.B i
Toggle the \fBinfo\fR mode. The info mode defines what hifs shows in the 
third column of the screen. The following info mode are available: username, 
process id, wchan, priority and command line.
.TP
.B m
Toggle the \fBmemory\fR mode. Hifs can show you the amount of free mem/swap 
or the amount of used mem/swap.
.TP
.B p
Select a process and set it's priority (nice value). You need enough 
priviliges to do this.
.TP
.B w
Write a message to the standard output of a process. You need enough 
priviliges to do this.
.TP
.B k
Select a process and kill it. First send a SIGHUP, then a SIGTERM and if 
the process is not gone by then, a SIGKILL is sent.
.TP
.B K
Select a process and kill it with a SIGKILL right away.
.TP
.B u
Set the period between two updates of the process data.
.TP
.B r
If configured, su to root. Another invoke drops the root priviliges.
.TP
.B CTRL-L
Redraw the screen.
.TP
.B q
Quit hifs.

.SH CONFIGFILE
In the hifs configfile, a setup can be defined. The name for this file is
\fB${HOME}/.hifsrc\fR. If this file is not found, reasonable default values 
are used.  The following directives can be used in the configfile: Every 
directive on one line. Lines beginning with a hash ('#') are ignored.

.TP
.B sort cpu|rss|vsize
Specify the sort mode. 
.TP
.B info username|pid|cmdline|wchan|priority
Specify the info mode.
.TP
.B memory free|used
Specify the memory mode.
.TP
.B delay DELAY
Specify the delay between updates in seconds. DELAY must be an int or a float.
.TP
.B diskfree SIZE
Specify the minimum free space in bytes per filesystem. Hifs notifies the 
user if the free space of a certain filesystem drops below this value. Only
read-write mounted filesystems of the type ext2, nfs and umsdos are checked. 
SIZE must be an int.
.TP
.B mapfile FILENAME
Specify the kernel symbol table. This file is generated during the compilation
of a kernel. By default, the following locations are searched in their 
respective order: /boot/System.map-%v, /boot/System.map, 
/lib/modules/%v/System.map and /usr/src/linux/System.map. %v Is the full
kernel version number. FILENAME must be a string.
.TP
.B group NAME { USER,ID USER,ID ... }
Define a group with name NAME. You can give up to eight USER, ID pairs. USER
is the login name of the user, ID is a single character that represents the 
user within hifs. There can be at most two groups. NAME and USER are strings.
ID is a char.

.SH NOTES
Hifs gets it's info from the \fBproc\fR-filesystem. It should be compiled 
into your kernel, and mounted on \fB/proc\fR. Without the /proc filesystem, 
hifs is useless and will refuse to run. 
.sp 0
Hifs is covered under the GNU General Public License (GPL).

.SH AUTHOR
The \fBLinux\fR version of Hifs is created by Geert Jansen (geertj@stack.nl). 
It is inspired by the original \fIFreeBSD\fR version, created by Sven Berkvens 
(sven@stack.nl). The latest verion can be found via anonymous ftp at 
skynet.stack.nl in /pub/linux/hifs.



CHANGES
Changes for hifs.

Changes from 1.3 to 1.4
-Bug fixes! 1.3 reached a big audience and with their bug reports I was able 
 to shake some stupid, some nifty bugs out.
-Added delay and diskfree option to the configfile.
-Added command line options.
-Don't wait for the user if a warning was issued unless in debug mode.
-Rewrote part of the manpage.
-Removed global configuration file. Only .hifsrc remains.
-Dynamically detect format of /proc/meminfo as the format keeps changing.
-In setuid version: prevent coredumps.
-Runtime shadow/normal password detection.
-Verbose banner.
-Better wchan support. Also try to read wchans from /boot/System.map and 
 some other locations: same behaviour as ps(1).
-Detection of possible too short delay between data updates.

Changes from 1.2 to 1.3
-Massive restructuring of the source. Things should be reasonably clean now.
-Only check selected fs-types for diskspace.
-New key bindings: I almost used up the entire alfabet, so I decided to do
 things differently. The sort/info/mem modes are now a toggle, so they need
 a total of only three keys. This way, hifs uses much less keys now.
-Hifs is more verbose now to the user.
-Changed utmp handling do go via libc and not via read(). This is cleaner.
-Added priority info mode.
-The configfile is now parsed via bison/flex. Extended syntax, like C-style
 strings in the configfile.
-Only one include file now: hifs.h. Everthing is included and defined here.
-New get_string() function that has some command-line editing and isn't as
 braindead as the ncurses mvinnstr() function.
-The update period can be real now, instead of only integer.
-Removed defaults.h.
-Hifs compiles cleanly under glibc-2. Fixed include files: don't include
 from asm/ and linux/ anymore.
-Changed to using autoconf instead of home-brew ./configure script.
-Added global configuration file. Search order is: $HOME/.hifsrc,
 $LIBDIR/hifs/system.hifsrc
-Added support for /proc/meminfo of kernel > 2.1.40.
-Runtime kernel version detection instead of compile time. This gives on hifs
 binary for all kernels.

Changes from 1.1 to 1.2
-Changed layout. First column is always process-name. Second is sorted on,
 and the third column can be freely chosen. This seems more intuitive.
-Added wchan info mode.
-Fixed "/proc is full" bug, when proc device is called something other than
 "none". (thanx: Masale).
-Pause when a warning is issued during initialisation.

Changes from first release to 1.1:
-Added verion number ...
-Added command-line info mode.
-Added delay option.
-Messaged are queued now, so only the most important messaged is shown.
-Added configure script, that looks for kernel and libc version.
-Removed System V IPC! 
-Removed fork(). Hifs is only one process now!
-Added (optional) setuid feature.
-Added more functionality to the "configure" script.
-Added renice feature.
-Added 'K' option: kill with SIGKILL right away.
-Added a configfile, for group, id, show and mem defaults (see manpage).
-100% is now 100%, instead of 100.8 %.

Version 1.0 didn't really have a version number, it was just "hifs". 
But I thought it would be easy after all, to have version numbers.


hifs.c
/* vi: ts=4 sw=4
 * 
 * Hifs -- Handy Information For Sysadmins
 * Copyright (C) 1996,1997 Geert Jansen 
 *  
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 *
 * The author can be contacted via email as <geertj@stack.nl>
 */

#include "hifs.h"

/* ------------------------------------------------------------------------
 * Globals. */

char *			tty 		= NULL;
struct stat		ttybuf;
int				warned		= 0;
int				rootflag	= 0;
char *			user		= NULL;

struct group * 	groups		= NULL;
int 			ngroups		= 0;
int				mlgroups	= 0;

char 			config_file[FILENAME_MAX];
int				version_code;
char *			version_string;

/* Defaults */

double			delay		= 5;
int				sort		= SORT_CPU;
int				memory		= MEM_FREE;
int				info		= INFO_NAME;
char *			mapfile		= "";

int				min_diskfree	= 1000000;
int				debug		= 0;

/* Tables for string representations of sort/info/mem modes */

struct mode sortmodes[] = {
	{ "By CPU", "CPU" },
	{ "By RSS", "RSS" },
	{ "By Vsize", "VSZ" }
};

struct mode infomodes[] = {
	{ "Process ID", "PID" },
	{ "Command Line", "CMD" },
	{ "Wchan", "WCH" },
	{ "Username", "NAM" },
	{ "Priority", "PRI" }
};

struct mode memmodes[] = {
	{ "Free (Kb)", "FRE" },
	{ "Used (Kb)", "USD" }
};


/* ------------------------------------------------------------------------
 * Prototypes not in hifs.h. */

void 		gracefull_exit	(int);
int			cfgfile			(void);
void		set_update		(double);
void		print_banner	(void);
void		print_help		(void);

/* ------------------------------------------------------------------------
 * cfgfile: Parse the configfile */

int cfgfile( void)
{
	char fname[FILENAME_MAX];
	int ret;
	struct passwd * pwd;

	extern FILE * yyin;
	extern int yyparse( void);

	if (!(pwd = getpwuid( getuid()))) {
		perror( "getpwuid()");
		return (1);
	}
	sprintf( fname, "%s/.hifsrc", pwd->pw_dir);

	/* If user does not have a .hifsrc, return */

	if (access( fname, F_OK))
		return (0);

	if (!(yyin = fopen( fname, "r"))) {
		perror( fname);
		return (1);
	}
	ret = yyparse();
	fclose( yyin);

	return (ret);
}

/* ------------------------------------------------------------------------
 * set_update: Set the update frequency. */

void set_update( double update)
{
	struct itimerval it;

	it.it_interval.tv_sec = update;
	it.it_interval.tv_usec = (update - floor( update)) * 1e6;
	it.it_value.tv_sec = update;
	it.it_value.tv_usec = (update - floor( update)) * 1e6;
	setitimer( ITIMER_REAL, &it, NULL);
}
	
/* ------------------------------------------------------------------------
 * gracefull_exit: This is the sig-go-away handler. */

void gracefull_exit( int sig)
{	
	proc_close();
	screen_close();
	printf( "\nExiting...(signal %d)\n", sig );
	chmod( tty, ttybuf.st_mode);
	_exit( 1);
}
	
/* ------------------------------------------------------------------------
 * print_banner: Show version information. */
 
void print_banner( void)
{
	printf( "Hifs %d.%d (%s@%s) (%s) %s\n", HIFS_MAJOR, HIFS_MINOR,
	        COMPILE_BY, COMPILE_HOST, COMPILE_CC, COMPILE_DATE);
	return;
}
	
/* ------------------------------------------------------------------------
 * print_help: Show some help about command line options. */

void print_help( void)
{
	printf( "Usage: hifs [OPTION]...\n\n");
	printf( "  -v, --version        show version information\n");
	printf( "  -h, --help           show this help\n");
	printf( "  -d, --debug          set debug mode\n");
	printf( "\n");
	return;
}

/* ------------------------------------------------------------------------
 * main: This is hifs. */

int main( int argc, char ** argv)
{
	char buf[BUFSIZ];
	char c, * ptr;
	int i, done, optindex; 
	int major, minor, patchlevel;
	struct sigaction sa;
	struct termios tioold, tionew;
	struct passwd * pwd;
	struct utsname ut;
	sigset_t sigset;

	struct option opts[] = {
		{ "version", 0, 0, 'v' },
		{ "help", 0, 0, 'h' },
		{ "debug", 0, 0, 'd'}
	};

#ifdef CONFIG_SU
	char * p, * pp, rootpass[32];
	struct rlimit rlim;

#ifdef CONFIG_SHADOW
	struct spwd * sp;
#endif

	/* Set core dump limit to 0 to prevent coredump of setuid binary.
	 * This may leave the encrypted root password visible in the core
	 * image. */

	if (getrlimit( RLIMIT_CORE, &rlim)) {
		perror( "getrlimit()");
		exit( 1);
	}
	rlim.rlim_cur = 0;
	if (setrlimit( RLIMIT_CORE, &rlim)) {
		perror( "setrlimit()");
		exit( 1);
	}

	/* Open the system log as `hifs' */

	openlog( "hifs", LOG_PID, LOG_USER);
	if (!(pwd = getpwuid( getuid()))) {
		syslog( LOG_WARNING, "started by uid %d not in password"
				" file!", getuid());
		exit( 1);
	}

	user = xmalloc( strlen( pwd->pw_name)+1);
	strcpy( user, pwd->pw_name);
	syslog( LOG_INFO, "started by %s (uid %d)", user, getuid());

	if (!(pwd = getpwuid( 0))) {
		perror( "getpwuid()");
		exit( 1);
	}
	strnzcpy( rootpass, pwd->pw_passwd, 32);

#ifdef CONFIG_SHADOW
	if ((sp = getspnam( pwd->pw_name)))
		strnzcpy( rootpass, sp->sp_pwdp, 32);
#endif

	/* Drop priviliges */

	if (seteuid( getuid())) {
		perror( "seteuid()");
		exit( 1);
	}

#else	/* ! CONFIG_SU */

	if (!(pwd = getpwuid( getuid()))) {
		perror( "getpwuid()");
		exit( 1);
	}
	user = xmalloc( strlen( pwd->pw_name)+1);
	strcpy( user, pwd->pw_name);

#endif  /* CONFIG_SU */

	/* Parse command-line arguments */
	
	while ((c = getopt_long( argc, argv, "vhd", opts, &optindex)) != EOF) {
		switch (c) {
		case 'v':
			print_banner();
			exit( 0);
		case 'h':
			print_help();
			exit( 0);
		case 'd':
			debug = 1;
			break;
		case '?':
			printf( "Try `hifs --help' for more information.\n");
			exit( 1);
		}
	}
	
	/* Get the kernel version */

	if (uname( &ut)) {
		perror( "uname()");
		exit( 1);
	}
	if (sscanf( ut.release, "%d.%d.%d", &major, &minor, &patchlevel) != 3) {
		fprintf( stderr, "kernel version: %s: unknown format\n", ut.release);
		exit( 1);
	}
	version_code = 1000000 * major + 1000 * minor + patchlevel;
	version_string = xmalloc( sizeof( ut.release) + 1);
	strcpy( version_string, ut.release);

	/* Parse the configfile */

	if (cfgfile()) {
		fprintf( stderr, "Failed to read the configfile\n");
		exit( 1);
	}
	
	if (proc_init()) {
		fprintf( stderr, "Initialisation of proc subsystem failed\n");
		fprintf( stderr, "Exiting...");
		exit( 1);
	}

	/* We make our tty mode 0600 to prevent talk's and write's to this
	 * window. */

	if (!(tty = ttyname(1))) {
		perror( "ttyname()");
		exit( 1);
	}	
	if (stat( tty, &ttybuf)) {
		perror( "stat()");
		warned = 1;
	}
	if (chmod( tty, 0600)) {
		perror( "chmod()");
		warned = 1;
	}

	/* If we printed a warning, wait for the user here if in debug mode. */

	if (warned && debug) {

		/* Save terminal attributes and go to cbreak mode. Then, wait
		 * for the user to press a key. */

		if (tcgetattr( STDIN_FILENO, &tioold)) {
			perror( "tcgetattr()");
			exit (1);
		}
		tionew = tioold;
		tionew.c_lflag &= ~(ICANON | ECHO);
		tionew.c_cc[VTIME] = 0;
		tionew.c_cc[VMIN] = 1;
		if (tcsetattr( STDIN_FILENO, TCSANOW, &tionew)) {
			perror( "tcsetattr()");
			exit (1);
		}
		printf( "\npress ANY key to continue");
		fflush( stdout);
		read( STDIN_FILENO, buf, 1);

		/* Reset terminal parameters */

		tcflush( STDIN_FILENO, TCIFLUSH);	/* discard more input */
		if (tcsetattr( STDIN_FILENO, TCSANOW, &tioold)) 
			perror( "tcsetattr()");
	}
	
	/* Initialise the screen */
	
	if (screen_init( 0)) {
		fprintf( stderr, "Initialisation of screen subsystem failed\n");
		fprintf( stderr, "Exiting...");
		exit( 1);
	}

	/* Initialise the cpu usage histories */

	for (i=1; i<5; i++) {
		proc_update( 0);
		screen_init( i);
		xsleep( 10);
	}
		
	/* Set up signal handlers.	*/

	sa.sa_handler = gracefull_exit;
	sigemptyset( &sa.sa_mask);
	sa.sa_flags = 0;
	sa.sa_restorer = NULL;
	sigaction( SIGHUP, &sa, NULL);
	sigaction( SIGINT, &sa, NULL);
	sigaction( SIGQUIT, &sa, NULL);
	sigaction( SIGTERM, &sa, NULL);
	sa.sa_handler = proc_update;
	sigaction( SIGALRM, &sa, NULL);

	/* Set updating frequency */

	set_update( delay);

	/* Create a signal set to block SIGALRM later on */

	sigemptyset( &sigset);
	sigaddset( &sigset, SIGALRM);

	/* Main program loop		*/

	done = 0;
	while (!done) {

		/* We block SIGALRM when screen_update() is called to prevent a data
		 * update while reading this data. */

		sigprocmask( SIG_BLOCK, &sigset, NULL);
		screen_update();
		sigprocmask( SIG_UNBLOCK, &sigset, NULL);

		switch (xgetch( BIG_SLEEP, 0)) {
			case 's': case ' ':
				sort++;
				if (sort > SORT_LAST)
					sort = 0;
				queue_msg( MIN_PRIO, "Sort mode: %s", sortmodes[sort].l);
				break;
			case 'i':
				info++;
				if (info > INFO_LAST)
					info = 0;
				queue_msg( MIN_PRIO, "Info mode: %s", infomodes[info].l);
				break;
			case 'm':
				memory++;
				if (memory > MEM_LAST)
					memory = 0;
				queue_msg( MIN_PRIO, "Mem mode: %s", memmodes[memory].l);
				break;
			case 'k':
				let_user_kill( KILL_NICE);
				break;
			case 'K':
				let_user_kill( KILL_BRUTE);
				break;
			case 'w':
				let_user_write();
				break;
			case 'p':
				let_user_renice();
				break;
			case 'u':
				ptr = get_string( "Update Period", 0);
				if (!ptr)
					break;
				if (sscanf( ptr, "%lf", &delay)) {
					set_update( delay);
					notice( "Update now %.2f s", delay);
				} else
					notice( "Illegal update");
				break;
			case 'h': case '?':
				show_help();	/* Fall tru */
			case CTRL('l'):
				screen_setup();
				break;

#ifdef CONFIG_SU
			case 'r':
				if (!geteuid()) {
					if (seteuid( getuid())) 
						notice( "Failed to drop privileges");
					else {
						syslog( LOG_INFO, "Priviliges dropped");
						notice( "Priviliges dropped");
						rootflag = 0;
					}
					break;
				}
				p = get_string( "pass", '*');
				if (!p)
					break;
				pp = crypt( p, rootpass);
				memset( p, 0, strlen( p));
				if (strcmp( pp, rootpass)) {
					syslog( LOG_WARNING, "Bad SU to root by %s (uid %d)",
							user, getuid());
					notice( "Illegal password!");
					break;
				}
				if (seteuid( 0)) 
					notice( "Failed to set priviliges");
				else {
					syslog( LOG_INFO, "SU to root by %s (uid %d)", 
							user, getuid());
					notice( "Hifs now running as root");
					rootflag = 1;
				}
				break;
#endif 	/* CONFIG_SU */

			case 'q':
				done = 1;
			default:
				break;
		}
	}
	
	proc_close();
	screen_close();
	chmod( tty, ttybuf.st_mode);

	return (0);
}

